### PR TITLE
Optimize loops

### PR DESCRIPTION
**Pull request by NRH-AA**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Remove redundant function calls in loops

**Issues addressed:** 

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by MillhioreBT** [comment]
I'm not very sure but I think that these special container methods are optimized by the compiler, I have not seen anywhere that it is a good practice or optimization method, but I think @ranisalt might have a quick and correct answer for this.

**Comment by NRH-AA** [comment]
At the very least its just good practice and there are many loops already in the source code that follow it. I figured its better to make it more consistent. As for if it matters for modern C++ I am in the same boat as you.  It could be that it is not redefining the size every iteration but if it isn't might as well follow best practice anyway.

**Comment by amatria** [comment]
> I'm not very sure but I think that these special container methods are optimized by the compiler

They probably are. This is good practice nonetheless: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

**Comment by MillhioreBT** [comment]
Honestly you should leave that as is, you're just adding unnecessary extra logic. the compiler already does this job and much better

Here I leave you an example to close this PR:
https://quick-bench.com/q/1TeRsXRYPbRehoc7AQCFj1H0snk

Even adding temporary variables to save the precalculated size is more time-consuming, it is negligible but...

<details>
<summary>Test image</summary>

image

</details>

**Comment by amatria** [comment]
> Even adding temporary variables to save the precalculated size is more time-consuming, it is negligible but...

I am sorry, but your benchmark is wrong. [Here](https://quick-bench.com/q/BY0l2olq81BEhZyvmEzvuGmrwws) are the correct results:

- `-O3`:

BY0l2olq81BEhZyvmEzvuGmrwws

- No optimizations:

TF4dOUi6aS20ivStSLwkoge-tNs

The version that stores the size of the vector is never slower than calling `size()` repeatedly.

**Comment by NRH-AA** [comment]
> Honestly you should leave that as is, you're just adding unnecessary extra logic. the compiler already does this job and much better
> 
> Here I leave you an example to close this PR: https://quick-bench.com/q/1TeRsXRYPbRehoc7AQCFj1H0snk
> 
> Even adding temporary variables to save the precalculated size is more time-consuming, it is negligible but...
> 
> Test image
> And if that were not enough, here is the comparison without compiler optimizations: https://quick-bench.com/q/2ZQg5xSsugJ-KwCjbdZHQCCy2gA image

Had to make a couple edits to the code you ran. I also increased the size of the arrays being checked and you can see the difference a little better. Please try with this.

`static void ContainerPrecalculateSize(benchmark::State& state) {
  std::vector<int> numbers = {};
  for (int i = 1; i < 50000; ++i) {
    numbers.push_back(i);
  }

  for (auto _ : state) {
    for (size_t i = 0, n = numbers.size(); i < n; i++) {
    }
  }
}
BENCHMARK(ContainerPrecalculateSize);

static void ContainerSize(benchmark::State& state) {
  std::vector<int> numbers = {};
  for (int i = 1; i < 50000; ++i) {
    numbers.push_back(i);
  }

  for (auto _ : state) {
    for (size_t i = 0; i < numbers.size(); i++) {
    }
  }
}
BENCHMARK(ContainerSize);`

**Comment by amatria** [comment]
@NRH-AA @MillhioreBT 

This benchmarking code is erroneous, resulting in the compiler optimizing away all meaningful computations. Please enable the "Show Noop bar" option. Refer to [my comment above](https://github.com/otland/forgottenserver/pull/4638#issuecomment-2027793844) for the correct benchmark.

File: src/iologindata.cpp [hunk 1]
```diff
@@ -687,7 +687,7 @@ bool IOLoginData::saveItems(const Player* player, const ItemBlockList& itemList,
 		}
 	}
 
-	for (size_t i = 0; i < containers.size(); i++) {
+	for (size_t i = 0, n = containers.size(); i < n; i++) {
```
Comment (ranisalt) [inline src/iologindata.cpp:690]: Note that this is _not_ equivalent; the size of `containers` is changed (in L700) so this change breaks the code.

**Comment by ranisalt** [comment]
One of these changes is not equivalent and makes the code incorrect; the rest, I think we should always write the simplest code, in that case it is keeping unchanged.

**Comment by NRH-AA** [comment]
I probably should of looked a little closer on the loops to make sure the sizes weren't changing during them. I do want to challenge the idea of simplest code and tfs though haha. Anyway, ill keep that in mind. Just trying to find places to help. Thanks for feedback.

**Pull request closed without merging**